File: labs/mobile/donor-archive/tunnel/PolicyRulesetRouter.kt
package com.luminet.android.tunnel

enum class AndroidPolicyVerdict {
    DIRECT, PROXY, REJECT
}

data class AndroidCidrRule(
    val netAddr: Long,
    val mask: Long,
    val verdict: AndroidPolicyVerdict
)

class PolicyRulesetRouter(private val defaultPolicy: AndroidPolicyVerdict = AndroidPolicyVerdict.DIRECT) {
    private val exactDomains = mutableMapOf<String, AndroidPolicyVerdict>()
    private val suffixDomains = mutableMapOf<String, AndroidPolicyVerdict>()
    private val keywordRules = mutableListOf<Pair<String, AndroidPolicyVerdict>>()
    private val cidrRules = mutableListOf<AndroidCidrRule>()
    private val cache = mutableMapOf<String, AndroidPolicyVerdict>()

    fun addExactDomain(domain: String, verdict: AndroidPolicyVerdict) {
        exactDomains[domain.trim().lowercase()] = verdict
    }

    fun addSuffixDomain(suffix: String, verdict: AndroidPolicyVerdict) {
        val clean = suffix.trim().lowercase().removePrefix(".")
        suffixDomains[clean] = verdict
    }

    fun addKeyword(keyword: String, verdict: AndroidPolicyVerdict) {
        keywordRules.add(keyword.trim().lowercase() to verdict)
    }

    fun addCidr(ipStr: String, maskBits: Int, verdict: AndroidPolicyVerdict) {
        val ipLong = parseIpv4ToLong(ipStr) ?: return
        val mask = if (maskBits == 0) 0L else (0xFFFFFFFFL shl (32 - maskBits)) and 0xFFFFFFFFL
        cidrRules.add(AndroidCidrRule(ipLong and mask, mask, verdict))
    }

    fun resolveDomain(domain: String): AndroidPolicyVerdict {
        val clean = domain.trim().lowercase()
        cache[clean]?.let { return it }

        // 1. Exact match
        exactDomains[clean]?.let {
            cache[clean] = it
            return it
        }

        // 2. Suffix match
        for ((suffix, v) in suffixDomains) {
            if (clean == suffix || clean.endsWith(".$suffix")) {
                cache[clean] = v
                return v
            }
        }

        // 3. Keyword match
        for ((kw, v) in keywordRules) {
            if (clean.contains(kw)) {
                cache[clean] = v
                return v
            }
        }

        cache[clean] = defaultPolicy
        return defaultPolicy
    }

    fun resolveIp(ipStr: String): AndroidPolicyVerdict {
        val ipLong = parseIpv4ToLong(ipStr) ?: return defaultPolicy
        for (c in cidrRules) {
            if ((ipLong and c.mask) == c.netAddr) {
                return c.verdict
            }
        }
        return defaultPolicy
    }

    fun clearCache() {
        cache.clear()
    }

    private fun parseIpv4ToLong(ip: String): Long? {
        val parts = ip.split(".")
        if (parts.size != 4) return null
        var res = 0L
        for (p in parts) {
            val octet = p.toLongOrNull() ?: return null
            if (octet !in 0..255) return null
            res = (res shl 8) or octet
        }
        return res
    }
}

File: labs/control-ui/api-shadow/policy_ruleset_router.ts
export type PolicyVerdict = 'DIRECT' | 'PROXY' | 'REJECT';

export interface CidrRuleEntry {
  netAddr: number;
  mask: number;
  verdict: PolicyVerdict;
}

export class PolicyRulesetRouter {
  private exactDomains = new Map<string, PolicyVerdict>();
  private suffixDomains = new Map<string, PolicyVerdict>();
  private keywordRules: Array<[string, PolicyVerdict]> = [];
  private cidrRules: CidrRuleEntry[] = [];
  private cache = new Map<string, PolicyVerdict>();
  public defaultPolicy: PolicyVerdict;
  constructor(defaultPolicy: PolicyVerdict = 'DIRECT') {
    this.defaultPolicy = defaultPolicy;
  }

  addExactDomain(domain: string, verdict: PolicyVerdict): void {
    this.exactDomains.set(domain.trim().toLowerCase(), verdict);
  }

  addSuffixDomain(suffix: string, verdict: PolicyVerdict): void {
    const clean = suffix.trim().toLowerCase().replace(/^\./, '');
    this.suffixDomains.set(clean, verdict);
  }

  addKeyword(keyword: string, verdict: PolicyVerdict): void {
    this.keywordRules.push([keyword.trim().toLowerCase(), verdict]);
  }

  addCidr(ipStr: string, maskBits: number, verdict: PolicyVerdict): void {
    const ipLong = this.ipToLong(ipStr);
    if (ipLong === null) return;
    const mask = maskBits === 0 ? 0 : (~0 << (32 - maskBits)) >>> 0;
    this.cidrRules.push({
      netAddr: (ipLong & mask) >>> 0,
      mask,
      verdict,
    });
  }

  resolveDomain(domain: string): PolicyVerdict {
    const clean = domain.trim().toLowerCase();
    const cached = this.cache.get(clean);
    if (cached) return cached;

    // 1. Exact match
    const exact = this.exactDomains.get(clean);
    if (exact) {
      this.cache.set(clean, exact);
      return exact;
    }

    // 2. Suffix match
    for (const [suffix, v] of this.suffixDomains.entries()) {
      if (clean === suffix || clean.endsWith(`.${suffix}`)) {
        this.cache.set(clean, v);
        return v;
      }
    }

    // 3. Keyword match
    for (const [kw, v] of this.keywordRules) {
      if (clean.includes(kw)) {
        this.cache.set(clean, v);
        return v;
      }
    }

    this.cache.set(clean, this.defaultPolicy);
    return this.defaultPolicy;
  }

  resolveIp(ipStr: string): PolicyVerdict {
    const ipLong = this.ipToLong(ipStr);
    if (ipLong === null) return this.defaultPolicy;
    for (const c of this.cidrRules) {
      if ((ipLong & c.mask) >>> 0 === c.netAddr) {
        return c.verdict;
      }
    }
    return this.defaultPolicy;
  }

  clearCache(): void {
    this.cache.clear();
  }

  private ipToLong(ip: string): number | null {
    const parts = ip.split('.').map((p) => parseInt(p, 10));
    if (parts.length !== 4 || parts.some((p) => isNaN(p) || p < 0 || p > 255)) {
      return null;
    }
    return ((parts[0]! << 24) | (parts[1]! << 16) | (parts[2]! << 8) | parts[3]!) >>> 0;
  }
}
